### ISSUE TITLE
Why can Pretrained Sentence Transformers models be used without sentence-transformers ?

### ISSUE THREAD

**Issue by qiulang** [UKPLab/sentence-transformers#2958]
I know little about sentence embedding models and I just want a decent sentence embedding to make my RAG work, so I am surprised to see that for **all** Pretrained Models listed in [sbert](https://sbert.net/docs/sentence_transformer/pretrained_models.html),  their HuggingFace pages _all_ mention these words, I use [all-MiniLM-L6-v2](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2) as an example, 

> Without [sentence-transformers](https://www.sbert.net/), you can use the model like this: First, you pass your input through the transformer model, then you have to apply the correct pooling-operation on-top of the contextualized word embeddings.

```
#Mean Pooling - Take average of all tokens
def mean_pooling(model_output, attention_mask):
    token_embeddings = model_output.last_hidden_state
    input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)
...

```

So why can these pretrained Sentence Transformers models be used without sentence-transformers ? And if I can HuggingFace Transformers directly, what are the relation between these model and sbet ? Does sbet just help me use those models easily? For example, what does mean_pooling do exactly ?

Adding to my confusion, I find BAAUI's [BEG models](https://huggingface.co/collections/BAAI/bge-66797a74476eb1f085c7446d) seem not to be the sentence-bert model, but they can also be used through sbet.

**Comment by tomaarsen** [comment]
Hello!

The [Structure of Sentence Transformer Models](https://sbert.net/docs/sentence_transformer/usage/custom_models.html#structure-of-sentence-transformer-models) documentation might be somewhat helpful here. The Transformer module described there is a thin wrapper around the Hugging Face Transformers `AutoModel`.
So, for example `all-MiniLM-L6-v2` consists of 3 modules, defined in the [modules.json](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/blob/main/modules.json) file: Transformer, Pooling, and Normalize. 

The snippet that you're referring to in your question is the torch-only implementation of the mean pooling that Sentence Transformers abstracts away in its Pooling module (note: the [1_Pooling/config.json](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/blob/main/1_Pooling/config.json) file determines the settings for the pooling strategy). To clarify a bit further: the Transformers AutoModel produces token embeddings, whereas we're interested in text embeddings instead. So, we use a pooling strategy to go from token embeddings -> text embeddings. Sentence Transformers just abstracts away this conversion (and e.g. also normalization), so that users don't have to worry about it.

An added bonus of this reliance on Transformers is that models not trained with Sentence Transformers, like the BGE models, can still be adapted to work in Sentence Transformers as long as the model 1) can be loaded with AutoModel and 2) produces token embeddings. 

Beyond that, Sentence Transformers supports a lot of [loss functions](https://sbert.net/docs/sentence_transformer/loss_overview.html) and a [Trainer](https://sbert.net/docs/sentence_transformer/training_overview.html) that can help with getting the best performing models for your task. As you can see in the [recent Sentence Transformer models](https://huggingface.co/models?library=sentence-transformers&sort=created), there's a lot of interest in finetuning. There might also already be some finetuned models that suit your use case, even if it's niche (e.g. 2 hours ago someone uploaded [this Turkish model](https://huggingface.co/alperctnkaya/bge-m3-distilled-en-tr) trained via distillation into bge-m3)

**In short:** Sentence Transformers is a wrapper around Transformers that 1) abstracts away common embedding model functionality such as pooling and normalization for easy 3-line usage (import; load; encode) and 2) implements a whole range of finetuning functionality that can help you get the best model on exactly your data. 

- Tom Aarsen

**Comment by qiulang** [comment]
Thanks for your detailed answer. Now I understand the process from token embeddings -> text embeddings. I still have some questions:

1. If I want to use `google-bert/bert-base-multilingual-cased` to do sentence embedding, what is the correct way to do that (Or is that a good idea)?  The sbert document you mentioned also use "bert-base-uncased" as an example. 

I googled it too and find this SO question[BERT sentence embeddings from transformers](https://stackoverflow.com/questions/63461262/bert-sentence-embeddings-from-transformers) 

Some answer there also used sbert, but its code seem a bit complex that the codes in sbert document.  Other answers suggested other way.

2. Does sbert have a specific document (other than the one you mentioned) to explain the relation between Sentence Transformer model and  Hugging Face Transformers  model ? I feel others may have this question too and your answer is very helpful.

**Comment by ir2718** [comment]
@qiulang 

> If I want to use google-bert/bert-base-multilingual-cased to do sentence embedding, what is the correct way to do that (Or is that a good idea)? The sbert [document](https://sbert.net/docs/sentence_transformer/usage/custom_models.html#sentence-transformer-model-from-a-transformers-model) you mentioned also use "bert-base-uncased" as an example.

You can use it, but it probably won't do very well. BERT is pretrained, but not fine-tuned for your particular use case. You can either fine-tune such a model for your use-case or use a different model which is already fine-tuned for eg. retrieval. The code you found is a bit different as it uses the previous version of the library.

> Does sbert have a specific document (other than the one you mentioned) to explain the relation between Sentence Transformer model and Hugging Face Transformers model ? I feel others may have this question too and your answer is very helpful.

As you said it yourself, most model cards provide instructions on how to use it with vanilla transformers. I would say the most "official" document where you can understand the relationship between the two is the original [SBERT paper](https://arxiv.org/abs/1908.10084), especially Figure 2. If memory serves me well, the previous docs included this particular figure somewhere. 

Although one thing that I found misleading from the figure is that the weights of the two BERTs are actually tied. You can also think of it as passing one sentence to a model, holding onto that embedding, passing the second sentence to the model, and comparing the two resulting embeddings.

**Comment by qiulang** [comment]
@ir2718 The reason I was thinking about using HuggingFace Transformer models directly (and my issue in the first place) is because I find [sbert Multilingual Models](https://www.sbert.net/docs/sentence_transformer/pretrained_models.html#multilingual-models) does not work well  for Chinese language in my test. So I am looking for other options and as I said in my first comment I know little about sentence embedding models so I am not able to fine tuning model at the moment. I am only able to find the available  sentence embedding models, compare them.

Then I have another thought that is it because sentence transformer does not necessarily work well for Chinese language? Other than sentence transformer, are there other sentence embeddings methods? 

I feel the information here [Performance Comparison](https://www.sbert.net/docs/pretrained-models/nli-models.html#performance-comparison) is a bit out-dated. For example, InferSent seems dead now and the link of [STS benchmark dataset](http://ixa2.si.ehu.es/stswiki/index.php/STSbenchmark) was broken.

**Comment by qiulang** [comment]
@tomaarsen @ir2718 your comments help me understand my question. Before I asked I was particularly confused why  BGE models can be loaded with sbert (as I am concerned more about Chinese sentence embedding). [BGE FAQ has these words](https://github.com/FlagOpen/FlagEmbedding/tree/master/FlagEmbedding/baai_general_embedding)

> Different from other embedding models using mean pooling, BGE uses the last hidden state of [cls] as the sentence embedding: sentence_embeddings = model_output[0][:, 0]. If you use mean pooling, there will be a significant decrease in performance.

Its [1_Pooling/config.json](https://huggingface.co/BAAI/bge-large-zh-v1.5/blob/main/)

```
{

     "word_embedding_dimension": 1024,
     "pooling_mode_cls_token": true,
     "pooling_mode_mean_tokens": false,
     "pooling_mode_max_tokens": false,
     "pooling_mode_mean_sqrt_len_tokens": false
 }
```

Without your explanation I have no idea what those words mean at all. So I am closing my question. Thanks again.
